presto-router/src/test/java/com/facebook/presto/router/TestSelectors.java

```diff
@@ -0,0 +1,210 @@
+import java.util.regex.Pattern;
+
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CLIENT_TAGS;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_SOURCE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_USER;
+import static com.facebook.presto.router.scheduler.SchedulerType.RANDOM_CHOICE;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.fail;
+
+public class TestSelectors
+{
+    private List<TestingPrestoServer> prestoServers;
+    private ClusterManager clusterManager;
+    private LifeCycleManager lifeCycleManager;
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        File configFile = File.createTempFile("router", "json");
+
+        Logging.initialize();
+
+        // set up server
+        ImmutableList.Builder<TestingPrestoServer> builder = ImmutableList.builder();
+        for (int i = 0; i < 3; ++i) {
+            TestingPrestoServer server = new TestingPrestoServer();
+            server.installPlugin(new TpchPlugin());
+            server.createCatalog("tpch", "tpch");
+            server.refreshNodes();
+            builder.add(server);
+        }
+
+        prestoServers = builder.build();
+
+        List<GroupSpec> groups = new ArrayList<>();
+        List<SelectorRuleSpec> selectors = new ArrayList<>();
+        Optional<SchedulerType> schedulerType = Optional.of(RANDOM_CHOICE);
+        Optional<URI> predictorUri = Optional.empty();
+
+        int i = 1;
+        for (TestingPrestoServer s : prestoServers) {
+            List<URI> members = new ArrayList<>();
+            members.add(URI.create(s.getBaseUrl().toString()));
+            GroupSpec groupSpec = new GroupSpec("group" + i, members, Optional.empty());
+            groups.add(groupSpec);
+
+            Optional<Pattern> sourceRegex = Optional.of(Pattern.compile("source" + i));
+            Optional<Pattern> userRegex = Optional.of(Pattern.compile("user" + i));
+
+            List<String> clientTagList = new ArrayList<>();
+            clientTagList.add("tag" + i);
+            Optional<List<String>> clientTags = Optional.of(clientTagList);
+
+            String targetGroup = "group" + i;
+            SelectorRuleSpec selectorRuleSpec = new SelectorRuleSpec(sourceRegex, userRegex, clientTags, targetGroup);
+            selectors.add(selectorRuleSpec);
+            i++;
+        }
+
+        RouterSpec routerSpec = new RouterSpec(groups, selectors, schedulerType, predictorUri);
+        JsonCodec<RouterSpec> codec = jsonCodec(RouterSpec.class);
+        String configTemplate = codec.toJson(routerSpec);
+
+        FileOutputStream fileOutputStream = new FileOutputStream(configFile);
+        fileOutputStream.write(configTemplate.getBytes(UTF_8));
+
+        fileOutputStream.close();
```

Comment:
nit: use a try-with-resources unless not possible. It will guarantee closing to free resources
```suggestion
        try (FileOutputStream fileOutputStream = new FileOutputStream(configFile)) {
            fileOutputStream.write(configTemplate.getBytes(UTF_8));
        }
```